examples/algorithms/tsne.py

```diff
@@ -0,0 +1,169 @@
+    for i in range(max_iter):
+        guess = (lower + upper) / 2.0
+        val = eval_fn(guess, i, sigma)
+        if val > target:
+            upper = guess
+        else:
+            lower = guess
+
+        if abs(val - target) <= tol:
+            break
+    return guess
+
+@phyfun
+def calc_perplexity(prob_matrix):
+    entropy = -sum( prob_matrix * log2(prob_matrix), axis=1)
+    perplexity = pow(2.0, entropy)
+    return entropy
+
+@phyfun
+def perplexity(distances, sigmas):
+    return calc_perplexity(calc_prob_matrix(distances, sigmas))
+
+@phyfun
+def eval_fn(distances, i, sigma):
+    return perplexity( distances[i:i+1, :], sigma )
+
+@phyfun
+def find_optimal_sigmas(distances, target_perplexity):
+    N = shape(distances, 0)
+    sigmas = array(0.0, N)
+    # TODO: parallelize this block?
+    #
+    for i in range( N ):
+        correct_sigma = bin_search(eval_fn, target_perplexity, i, sigma)
+        sigmas = vstack(sigmas, correct_sigma)
+    return sigmas
+
+@phyfun
+def p_conditional_to_joint(P):
+    return (P + transpose(P)) / (2.0 * float(shape(P, 0)))
+
+# NOTE: Ignore these for now...
+#
+#@phyfun
+#def q_joint(Y):
+#    dists = neg_squared_euc_dists(Y)
+#    exp_dists = exp(distances)
+#    # TODO: does fill return a ref to inv_distances?
+#    #
+#    fill_diagonal(exp_dists, 0.0)
+#
+#    # TODO: return tuples?
+#    #
+#    return exp_dists / sum(exp_dists), None
+#
+#@phyfun
+#def symmetric_sne_grad(P, Q, Y):
+#    pq_diff = P-Q
+#    pq_expanded = expand_dims(pq_diff, 2) # NxNx1
+#    y_diffs = expand_dims(Y, 1) - expand_dims(Y, 0) # NxNx2
+#    grad = 4.0 * sum(pq_expanded * y_diffs, axis=1) # Nx2
+#    return grad
+
+@phyfun
+def q_tsne(Y):
+    distances = neg_squared_euc_dists(Y)
+    inv_distances = pow(1.0-distances, -1.0)
+    # TODO: does fill return a ref to inv_distnaces?
+    #
+    fill_diagonal(inv_distances, 0.0)
+
+    # TODO: return tuples?
+    #
+    return inv_distances / sum(inv_distances), inv_distances
+
+@phyfun
+def tsne_grad(P, Q, Y, distances):
+    pq_diff = P - Q
+    pq_expanded = expand_dims(pq_diff, 2) # NxNx1
+    y_diffs = expanded_dims(Y, 1) - expanded_dims(Y, 0) # NxNx2
+    distances_expanded = expand_dims(distances, 2) # NxNx1
+    y_diffs_wt = y_diffs * distances_expanded # NxNx2
+    grad = 4.0 * sum(pq_expanded * y_diffs_wt, axis=1) # Nx2
+    return grad
+
+@phyfun
+def p_joint(X, target_preplexity):
+    distances = neg_squared_euc_dists(X)
+    sigmas = find_optimal_sigmas(distances, target_perplexity)
+    p_conditional = calc_prob_matrix(distances, sigmas)
+    P = p_conditional_joint(p_conditional)
+    return P
+
+@phyfun
+def estimate_sne(X, y, P, rng, num_iters, q_fn, grad_fn, learning_rate, momentum):
+    shape_x_arr = array(0, 2)
+    shape_x_arr[0] = shape(X, 0)
+    shape_x_arr[1] = 2
+
+    # TODO: https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.RandomState.html
+    Y = random('normal',(0.0, 0.0001, shape_x_arr)
```

Comment:
trying to emulate this capability: https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.RandomState.html

Reply:
FWIW, The existing `random()` primitive relies on balze's `rand()` generators which in turn rely in `std::uniform_int_distribution` and `std::uniform_real_distribution` of numbers generated by the standard Mersenne Twister (`mt19937`) generator. We will need to add our own facilities compatible/similar to those in blaze to be able to use other distributions (like 'normal'), see #207.

Reply:
See also https://bitbucket.org/blaze-lib/blaze/issues/158/random-number-distributions.